In [1]:
import numpy as np
import pandas as pd

In [5]:
df = pd.read_csv('../data/movies_metadata.csv')
df.head(3)

/tmp/ipykernel_158263/2807870231.py:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/movies_metadata.csv')


adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   

     original_title                                           overview  ...  \
0         Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1           Jumanji  When siblings Judy and Peter discover an encha...  ...   
2  Grumpier Old Men  A family wedding reignites the ancient feud be...  ...   

  release_date      revenue runtime  \
0   1995-10-30  373554033.0    81.0   
1   1995-12-15  262797249.0   104.0   
2   1995-12-22          0.0   101.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

  vote_average vote_count  
0          7.7     5415.0  
1          6.9     2413.0  
2          6.5       92.0  

[3 rows x 24 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [6]:
m = df.vote_count.quantile(0.80)
print(m)

50.0


- It means that only 20% of the movies has gathered more than 50 votes

The selection criteria for the movies are:
- consider only those movies that have gathered more than 50 votes
- movies have runtime greater than 45 minutes and less than 300 minutes

In [7]:
q_movies = df[(df["runtime"] >= 45) & (df["runtime"] <= 300)]
q_movies = q_movies[q_movies['vote_count'] >= m]
q_movies.shape

(8963, 24)

- Only 8963 movies satisfy the above two conditions out of 45466 movies

Weighted Rating(WR) = (v / (v + m)) * R + (m / (v + m)) * C

where,
- v is the number of votes garnered by the movie
- m is the minimum number of the votes required for the movie to be in the chart
- R is the mean rating of the moive
- C is the mean rating of all the movies in the dataset

In [19]:
C = df['vote_average'].mean()
print(C)

5.618207215134185


In [10]:
def weighted_rating(x, m=m, c=C):
    v = x['vote_count']
    R = x['vote_average']
    return (v / (v + m) * R) + (m / (v + m) * C)


In [15]:
q_movies['score'] = q_movies.apply(weighted_rating, axis = 1)

In [16]:
q_movies.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
5  False                                                NaN  60000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
4                     [{'id': 35, 'name': 'Comedy'}]   
5  [{'id': 28, 'name': 'Action'}, {'id': 80, 'nam...   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
4                                   NaN  11862  tt0113041                en   
5                                   NaN    949  tt0113277                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
4  Father of the Bride Part II   
5                         Heat   

                                            overview  ...      revenue  \
0  Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1  When siblings Judy and Peter discover an encha...  ...  262797249.0   
2  A family wedding reignites the ancient feud be...  ...          0.0   
4  Just when George Banks has recovered from his ...  ...   76578911.0   
5  Obsessive master thief, Neil McCauley leads a ...  ...  187436818.0   

  runtime                                   spoken_languages    status  \
0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2   101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
4   106.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
5   170.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
4  Just When His World Is Back To Normal... He's ...   
5                           A Los Angeles Crime Saga   

                         title  video vote_average vote_count     score  
0                    Toy Story  False          7.7     5415.0  7.680953  
1                      Jumanji  False          6.9     2413.0  6.873979  
2             Grumpier Old Men  False          6.5       92.0  6.189510  
4  Father of the Bride Part II  False          5.7      173.0  5.681661  
5                         Heat  False          7.7     1886.0  7.646235  

[5 rows x 25 columns]

In [18]:
# sor the movies on the basis of weighted rating score
q_movies.sort_values('score', ascending=False).iloc[:25, :]

adult                              belongs_to_collection     budget  \
10309  False                                                NaN   13200000   
314    False                                                NaN   25000000   
834    False  {'id': 230, 'name': 'The Godfather Collection'...    6000000   
40251  False                                                NaN          0   
12481  False  {'id': 263, 'name': 'The Dark Knight Collectio...  185000000   
2843   False                                                NaN   63000000   
292    False                                                NaN    8000000   
522    False                                                NaN   22000000   
23673  False                                                NaN    3300000   
5481   False                                                NaN   15000000   
2211   False                                                NaN   20000000   
1178   False  {'id': 230, 'name': 'The Godfather Collection'...   13000000   
1152   False                                                NaN    3000000   
1176   False  {'id': 119674, 'name': 'Psycho Collection', 'p...     806948   
351    False                                                NaN   55000000   
1184   False                                                NaN   30000000   
1154   False  {'id': 10, 'name': 'Star Wars Collection', 'po...   18000000   
18465  False                                                NaN   13000000   
289    False                                                NaN   16000000   
3030   False                                                NaN   60000000   
1170   False                                                NaN   25000000   
2216   False                                                NaN   20000000   
1161   False                                                NaN     350000   
9698   False                                                NaN   24000000   
2884   False                                                NaN   26500000   

                                                  genres  \
10309  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
314    [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...   
834    [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...   
40251  [{'id': 10749, 'name': 'Romance'}, {'id': 16, ...   
12481  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...   
2843                       [{'id': 18, 'name': 'Drama'}]   
292    [{'id': 53, 'name': 'Thriller'}, {'id': 80, 'n...   
522    [{'id': 18, 'name': 'Drama'}, {'id': 36, 'name...   
23673                      [{'id': 18, 'name': 'Drama'}]   
5481   [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
2211   [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
1178   [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...   
1152                       [{'id': 18, 'name': 'Drama'}]   
1176   [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   
351    [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
1184   [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...   
1154   [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
18465  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...   
289    [{'id': 53, 'name': 'Thriller'}, {'id': 80, 'n...   
3030   [{'id': 14, 'name': 'Fantasy'}, {'id': 18, 'na...   
1170   [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...   
2216                       [{'id': 18, 'name': 'Drama'}]   
1161                       [{'id': 18, 'name': 'Drama'}]   
9698   [{'id': 14, 'name': 'Fantasy'}, {'id': 16, 'na...   
2884   [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                                                homepage      id    imdb_id  \
10309                                                NaN   19404  tt0112870   
314                                                  NaN     278  tt0111161   
834                         http://www.thegodfather.com/     238  tt0068646   
40251    https://www.funimationfilms.com/movie/yourname/  372058  tt5311514   
12481       http://the

we can observe that the movie **Dilwale Dulhania Le Jayenge** has the highest score but relatively few number of votes